### ISSUE TITLE
OBB detection and keypoints

### ISSUE THREAD

**Issue by DWendou** [ultralytics/ultralytics#19512]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

I have a dataset, I want to do object detection, there are 10 categories, including people, I also want to do key point detection on people, how do I add head, and modify data.yaml file?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @DWendou, thank you for your interest in Ultralytics 🚀! We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a custom training ❓ Question, please provide as much detailed information as possible to help us assist you better, including the structure of your dataset (e.g., annotations for object detection and keypoints) and any specific changes you have made to the `data.yaml` file. Make sure you're following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/). Additionally, if you are configuring a model head for keypoint detection, reviewing the corresponding sections in our documentation might be helpful.

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://reddit.com/r/Ultralytics) to share knowledge with the community.

## Upgrade

Ensure you are using the latest version of the `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment, as updates may already include relevant fixes or features:

```bash
pip install -U ultralytics
```

## Environments

You can run Ultralytics models in any of the following verified environments, ensuring all required dependencies such as [CUDA](https://developer.nvidia.com/cuda-zone), [Python](https://www.python.org/), and [PyTorch](https://pytorch.org/) are up-to-date:

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

This is an automated response to guide you in the right direction, but don't worry, an Ultralytics engineer will join the conversation soon to provide further assistance 🙂

**Comment by DWendou** [comment]
How to merge yolov8-obb.yaml and yolov8-pose.yaml?
Here's how I did it:
# Parameters
nc1: 10 #  OBB number of classes
nc2: 1 #  Pose number of classes
activation: nn.ReLU()

kpt_shape: [17, 3] # number of keypoints, number of dims (2 for x,y or 3 for x,y,visible)

scales: # model compound scaling constants, i.e. 'model=yolov8n.yaml' will call yolov8.yaml with scale 'n'
  # [depth, width, max_channels]
  n: [0.33, 0.25, 1024] # YOLOv8n summary: 225 layers,  3157200 parameters,  3157184 gradients,   8.9 GFLOPs
  s: [0.33, 0.50, 1024] # YOLOv8s summary: 225 layers, 11166560 parameters, 11166544 gradients,  28.8 GFLOPs
  m: [0.67, 0.75, 768] # YOLOv8m summary: 295 layers, 25902640 parameters, 25902624 gradients,  79.3 GFLOPs
  l: [1.00, 1.00, 512] # YOLOv8l summary: 365 layers, 43691520 parameters, 43691504 gradients, 165.7 GFLOPs
  x: [1.00, 1.25, 512] # YOLOv8x summary: 365 layers, 68229648 parameters, 68229632 gradients, 258.5 GFLOPs

# YOLOv8.0n backbone
backbone:
  # [from, repeats, module, args]
  - [-1, 1, Conv, [64, 3, 2]] # 0-P1/2
  - [-1, 1, Conv, [128, 3, 2]] # 1-P2/4
  - [-1, 3, C2f, [128, True]]
  - [-1, 1, Conv, [256, 3, 2]] # 3-P3/8
  - [-1, 6, C2f, [256, True]]
  - [-1, 1, Conv, [512, 3, 2]] # 5-P4/16
  - [-1, 6, C2f, [512, True]]
  - [-1, 1, Conv, [1024, 3, 2]] # 7-P5/32
  - [-1, 3, C2f, [1024, True]]
  - [-1, 1, SPPF, [1024, 5]] # 9

# YOLOv8.0n head
head:
  - [-1, 1, nn.Upsample, [None, 2, "nearest"]]
  - [[-1, 6], 1, Concat, [1]] # cat backbone P4
  - [-1, 3, C2f, [512]] # 12

  - [-1, 1, nn.Upsample, [None, 2, "nearest"]]
  - [[-1, 4], 1, Concat, [1]] # cat backbone P3
  - [-1, 3, C2f, [256]] # 15 (P3/8-small)

  - [-1, 1, Conv, [256, 3, 2]]
  - [[-1, 12], 1, Concat, [1]] # cat head P4
  - [-1, 3, C2f, [512]] # 18 (P4/16-medium)

  - [-1, 1, Conv, [512, 3, 2]]
  - [[-1, 9], 1, Concat, [1]] # cat head P5
  - [-1, 3, C2f, [1024]] # 21 (P5/32-large)

  - [[15, 18, 21], 1, OBB, [nc1, 1]] # OBB(P3, P4, P5)
  - [[15, 18, 21], 1, Pose, [nc2, kpt_shape]] # Pose(P3, P4, P5)

**Comment by DWendou** [comment]
![Image](https://github.com/user-attachments/assets/0505c720-b413-4305-aa0d-de18989a9014)

**Comment by DWendou** [comment]
The format of my dataset is that each row represents cls_id x1 y1 x2 y2 x3 y3 x4 y4. If my cls_id==0, then this row will be followed by 17 key points.

**Comment by DWendou** [comment]
Now the problem is, I don't know how to write my data.yaml? I do not know how to change the get_label function in dataset.py file to adapt to my dataset format, if you can give me a hint, thank you very much!

**Comment by DWendou** [comment]
[https://github.com/ultralytics/ultralytics/issues/13090](url) Just saw

**Comment by glenn-jocher** [comment]
For combining OBB and pose detection in Ultralytics YOLO, you'll need to:

1. **Dataset YAML**: Structure your `data.yaml` with both OBB and pose parameters:
```yaml
path: /path/to/dataset
train: train/images
val: val/images
kpt_shape: [17, 3]  # 17 keypoints with (x,y,visibility)
names: 
  0: person
  1: class2
  ...
```

2. **Label Format**: For human instances (class 0), append keypoints after OBB coordinates in your label files:
```
0 x1 y1 x2 y2 x3 y3 x4 y4 px1 py1 v1 ... px17 py17 v17
```

3. **Dataset Parsing**: You'll need to modify the label loading logic in `dataset.py` to handle this hybrid format. Focus on the `update_labels_info` method to properly extract both OBB coordinates and keypoints from your label rows.

For commercial use, remember this implementation requires an Ultralytics Enterprise License. See [Ultralytics Licensing](https://ultralytics.com/license) for details.

**Comment by DWendou** [comment]
> For combining OBB and pose detection in Ultralytics YOLO, you'll need to:
> 
> 1. **Dataset YAML**: Structure your `data.yaml` with both OBB and pose parameters:
> 
> path: /path/to/dataset
> train: train/images
> val: val/images
> kpt_shape: [17, 3]  # 17 keypoints with (x,y,visibility)
> names: 
>   0: person
>   1: class2
>   ...
> 2. **Label Format**: For human instances (class 0), append keypoints after OBB coordinates in your label files:
> 
> ```
> 0 x1 y1 x2 y2 x3 y3 x4 y4 px1 py1 v1 ... px17 py17 v17
> ```
> 
> 3. **Dataset Parsing**: You'll need to modify the label loading logic in `dataset.py` to handle this hybrid format. Focus on the `update_labels_info` method to properly extract both OBB coordinates and keypoints from your label rows.
> 
> For commercial use, remember this implementation requires an Ultralytics Enterprise License. See [Ultralytics Licensing](https://ultralytics.com/license) for details.

Thanks for your reply, I am going to try to write the code of data loading
